File: AnalisisNumerico/PivoteoParcial.java
package AnalisisNumerico;
public class PivoteoParcial {
	static double A[][] = {{2,-3,4,1},{-4,2,1,-2},{1,3,-5,3},{-3,-1,1,-1}};
	static double B[] = {10,-10,32,-21};
	static double AuxB[] = {10,-10,32,-21};
	static double cambios[][] = {{0},{1},{2},{3}};
	static int d = B.length;
	private static double[][] cambioL(){
		double L[][] = new double[d][d];
		double AuxL[][] = new double[d][d];
		cambios = Parcial(A,B,d,3);
		L = Parcial(A,B,d,0);
		AuxL = Parcial(A,B,d,0);
		for(int i=1;i<d;i++) {
			for(int j=0;j<d;j++) {
				//L[i][j] = AuxL[(int)cambios[j][i]][i];
				//System.out.println("Estoy en: "+i+" y "+j+" con: "+L[i][j]);
				//L[i][j] = AuxL[(int)cambios[i][j]][j];
				L[i][i]=0;
				//System.out.println("Estoy en: "+i+" y "+j+" con: "+L[i][j]);
			}
		}
		return L;
	}
	private static double[] cambiosB() {
		cambios = Parcial(A,B,d,3);
		for(int i=0;i<d;i++) {
			//B[i] = AuxB[(int)cambios[i][i]];
		}
		return B;
	}
	public static void main(String[] args) {
		//P=0 L, P=1 U, P=2 c, p>2 cambios
		Parcial(A,B,d,1);
		//SustiRegre.sustiRegre(Parcial(A,B,d,1),SustiProgre.sustiProgre(cambioL(), cambiosB(), d),d);
		}
	private static double[][] Parcial(double A[][], double B[], int d,int p) {
		int filam = 0;
		double ma,m = 0;
		double Aux[] = new double[d+1];
		double camfila[] = new double[d+1];
		double L[][] = new double[d][d];
		double U[][] = new double[d+1][d+1];
		double c[][] = new double[d+1][d+1];
		for (int i=0;i<=d;i++) {
			for (int j=0;j<d;j++) {
				if(i == d) {
					c[j][i] = B[j];
				}else {
					c[i][j]=A[i][j];
				}
			}
		}
		for (int k=0;k<d;k++) {
			filam = (int) BuscarMayor(c,d,k,0);
			if(filam!=k && k!=d) {
				ma =  BuscarMayor(c,d,k,1);
				if(ma==0) {
					System.out.println("Hay solo ceros en la columna: "+k);
					System.out.println("No se pudo completar el proceso");
				}else {
					c = cambioFila(c,k,filam);
					System.out.println("Soy el mayor: "+ma+" Etapa#: "+k);
					System.out.println("soy la fila donde estoy: "+filam);
				}
			}
			for(int i=k+1;i<d+1;i++) {
				if(c[k][k]==0){
					System.out.println("Hay un cero en la diagonal");
					System.out.println("No se pudo completar el proceso");
					k=d+1;
					break;
				}else {
					m = c[i][k]/c[k][k];
				}
				for(int j=0;j<=d;j++) {
					c[i][j]= c[i][j] - (m*c[k][j]);
				}
			}
		}if(p==0) {
			return L;
			}else if(p==1) {
				return U;
				}else{
					return c;
				}
	}
	private static double[][] cambioFila(double[][] c, int k, int filam) {
		double Mc[][] = new double[c.length][c.length];
		double Cam[] = new double[c.length];
		for(int i=0;i<c.length;i++) {
			Cam[i] = c[k][i];
		}
		for(int i = k;i<c.length;i++) {
			c[k][i] = c[filam][i];
			c[filam][i] = Cam[i];
		}
		for(int i=0;i<c.length;i++) {
			for(int j = 0;j<c.length;j++) {
				Mc[i][j] = c[i][j];
			}
		}
		return Mc;
	}
	private static double BuscarMayor(double A[][],int d, int k, int p) {
		double ma=0;
		int filam = 0;
		for(int j = k;j<d;j++) {
			if(Math.abs(A[j][k])>Math.abs(ma)) {
				ma = A[j][k];
				filam = j;
			}
		}
		if(p==0) {
			return filam;
		}else {
			return ma;
		}
	}
}

File: AnalisisNumerico/Sistema de ecuaciones lineales/PivoteoParcial.m
function PivoteoParcial(A,b)
  #Funciona con cualquier matriz A nxn y b nx1
  #Se solicita el valor p donde 0 es L
  #para p=1 es U y para P>1 || p<0 es b modificada. 
  disp("0) para la matriz L")
  disp("1) para la Matriz U")
  disp("2) para el vector b luego del ejecutar el método") 
  p = input("Elija una opción, por favor: ");
  while(p<0 || p>2)
    disp("No hay ingresado un número correctamente");
    p = input("Ingrese un número, por favor: ");
  endwhile
  A
  b
  c = [A, b];
  Filam = 0; %Fila donde se encuentra el mayor.
  ma = 0; m = 0;%ma es el mayor, y m es la multiplicidad por etapas.
  d = length(b);
  Auxi = 1:d;
  camfila = Auxi';
  AuxcamFila = Auxi';
  L = zeros(d);
  U = zeros(d);
  AuxA = A;
  for k=1:d
    A = c;
    [ma, Filam] = BuscarMayor(A,d,k);
    if Filam!=k && k!=d
      if ma==0
        disp('No se pudo completar el proceso')
        break;
      else
      camfila(k) = camfila(Filam);
      camfila(Filam) = AuxcamFila(k);
      AuxcamFila(Filam) = AuxcamFila(k);
      c = cambioFil(c,k,Filam); 
      endif
    endif
      for i=k+1:d
        if c(k,k) == 0
          disp('No se pudo completar el proceso')  
          break;
        else
          m = c(i,k)/c(k,k);%multiplicadores
        endif
        if i<d+1
          L(i,k) = m;%Matrix dxd con los multiplicadores
        endif
        for j=1:d+1
          c(i,j) = c(i,j) - m*c(k,j);
        endfor
      endfor
  endfor
  for i=1:d
    U(:,i) = c(:,i);
  endfor
  L = cambioL(AuxA,U,camfila);
  for i=1:d
    L(1,i) = 0;
    L(i,d) = 0;
  endfor
  bm = cambioB(b,camfila);
  z = zeros(d,1);
  z = SustiProgre(L,bm);
  if p==0
    L
    disp("x = ")
    SustiRegre(U,z);
  elseif p==1
    U
    disp("x = ")
    SustiRegre(U,z);
  else
    bm
    disp("x = ")
    SustiRegre(U,z);
  endif
 function [ma, Filam] = BuscarMayor(A,d,k)
   ma = 0;
   for j=k:d
     if abs(A(j,k))>abs(ma)
       ma = A(j,k);
       Filam = j;
     endif
   endfor
 endfunction
  function AuxC = cambioFil(c,k,Filam)
    AuxC = c;
    AuxC(k,:) = AuxC(Filam,:);
    AuxC(Filam,:) = c(k,:);
 endfunction
 function L = cambioL(A,U,camfila)
   AuxA = A;
   AuxL = zeros(length(camfila));
   for i=1:length(camfila)
     A(i,:)= AuxA(camfila(i),:);
     AuxA(camfila(i),:) = A(i,:);
   endfor
   xn = 0;
   for k=1:length(camfila)
     for i=1:length(camfila)
       xn = 0;
       for j=1:length(camfila)
         xn = xn + AuxL(i,j)*U(j,k);
       endfor
       L(i,k) = (A(i,k)-xn)/U(k,k);
       AuxL(i,k) = L(i,k);
       L(k,k) = 0;
     endfor
   endfor
 endfunction
 function Auxb = cambioB(b,camfila)
   Auxb = b;
   for i=1:length(camfila)
     Auxb(i) = b(camfila(i));
     b(camfila(i)) = Auxb(i);
   endfor
 endfunction
endfunction